ArjunaCore/arjuna/tests/classes/com/hp/mwtests/ts/arjuna/recovery/RecoverySuspendTest.java

```diff
@@ -0,0 +1,159 @@
+        };
+
+        // Let's go quick on this
+        recoveryConfig.setRecoveryBackoffPeriod(recoveryBackoffPeriod);
+        recoveryConfig.setPeriodicRecoveryPeriod(periodicRecoveryPeriod);
+        // don't sign off until the store is empty
+        recoveryConfig.setWaitForFinalRecovery(true);
+
+        // the test set of modules
+        recoveryConfig.setRecoveryModuleClassNames(Arrays.asList(modules));
+
+        // obtain a new RecoveryManager with the above config:
+        _manager = RecoveryManager.manager(RecoveryManager.INDIRECT_MANAGEMENT);
+        // recovery can start
+        _manager.startRecoveryManagerThread();
+
+        _recordTypeMap = new RecordTypeMap() {
+            @SuppressWarnings("unchecked")
+            public Class getRecordClass ()
+            {
+                return BytemanControlledRecord.class;
+            }
+
+            public int getType ()
+            {
+                return RecordType.USER_DEF_FIRST0;
+            }
+        };
+
+        RecordTypeManager.manager().add(_recordTypeMap);
+    }
+
+    @BeforeEach
+    public void beforeEach() {
+        // In case the Recovery Manager was suspended, this will resume it
+        _manager.resume();
+    }
+
+    @AfterAll
+    public static void afterClass() {
+        RecordTypeManager.manager().remove(_recordTypeMap);
+    }
+
+    @AfterEach
+    public void afterEach() {
+        if (!TxControl.isEnabled()) {
+            // Re-activate transaction creation
+            TxControl.enable();
+        }
+    }
+
+    @Test
+    public void testSuspensionWheneThereArentTxnsToRecover() {
+
+        // The Transaction System needs to be disabled, i.e. no new transactions can be created
+        TxControl.disable();
+
+        // Makes sure that the transaction reaper completed all transactions
+        TransactionReaper.transactionReaper().waitForAllTxnsToTerminate();
+
+        long startSuspension = System.currentTimeMillis();
+
+        // Synchronization is not needed (i.e. async = true) as isWaitForFinalRecovery() == true
+        _manager.suspend(true);
+        long durationOfSuspension = System.currentTimeMillis() - startSuspension;
+
+        Assertions.assertTrue(durationOfSuspension <
+                // In milliseconds
+                1000L * (periodicRecoveryPeriod + recoveryBackoffPeriod), String.valueOf(durationOfSuspension));
+    }
+
+    @Test
+    public void testSuspensionWhenThereIsAtomicActionToRecover() {
+
+        // Make sure that the test environment is ready
+        BytemanControlledRecord.resetCommitCallCounter();
+        BytemanControlledRecord.resetGreenFlag();
+        
+        createBasicAtomicAction();
+
+        // The Transaction System needs to be disabled, i.e. no new transactions can be created
+        TxControl.disable();
+
+        // Makes sure that the transaction reaper completed all transactions
+        TransactionReaper.transactionReaper().waitForAllTxnsToTerminate();
```

Comment:
I think this maybe should be `TransactionReaper.terminate(true)` (like in https://github.com/jmfinelli/narayana/pull/4/files#diff-7990bacf7d203544a582162b855619d55020e7bad28d66b6081822363250bec8R67)

Reply:
Actually, will use this thread given my previous comment just now :)

Reply:
Is that definitely the intent? I feel like if we have shutdown the transaction manager system (not the recovery manager system) and in-flight transactions are finished the reaper can (should?) be shutdown.

By the way, I can't recall precisely off the top of my head but is it possible that some transactions are in-flight that don't have a reaper element. If so, I expect we should be checking for those too.

Reply:
> Is that definitely the intent? I feel like if we have shutdown the transaction manager system (not the recovery manager system) and in-flight transactions are finished the reaper can (should?) be shutdown.

I see your point. What I was thinking when I wrote the code (and the above comment) is that when we shut down the Transaction System (i.e. `TxControl.disable`), what we actually do is avoiding that user threads can create transactions. OTOH, when we terminate the Transaction Reaper we actually terminate the embedded threads that handle/monitor in-flight transactions. That is why I think that the two operations (i.e. `TxControl.disable` and `TransactionReaper.terminate`) are semantically different, which is also supported by the existence of `TxControl.enable` and the absence of `TransactionReaper.enable` (TransactionReaper only has `TransactionReaper.instantiate`, which seems to imply that we should invoke that method when initialising Arjuna and not when resuming it)

> By the way, I can't recall precisely off the top of my head but is it possible that some transactions are in-flight that don't have a reaper element. If so, I expect we should be checking for those too.

I recall that the Transaction Reaper always monitors all in-flight transactions with a timeout. Were you thinking about transactions without a timeout? In this case, I think Recovery Modules implementing `SuspendBlockingRecoveryModule` should be responsible for handling those transactions as well. I will double check if this PoC satisfies this assumption. BTW, a test demonstrating this case will be very helpful :+1: Thanks Tom

Reply:
Thank you!

The discussion about the TransactionReaper suspend and resume is interesting to consider more. I am in favour of shutting it down and not refactoring the waiting outside of the shutdown method. I think I will discuss that in the thread where you made the change though. Sorry to move around a bit

Reply:
Related thread already resolved